src/fix/main.c

```diff
@@ -65,535 +75,944 @@ static void print_usage(void)
+{
+	va_list ap;
 
-	/*
-	 * Open the ROM file
-	 */
-
-	rom = fopen(argv[argc - 1], "rb+");
-
-	if (rom == NULL)
-		err(1, "Error opening file %s", argv[argc - 1]);
-
-	/*
-	 * Read ROM header
-	 *
-	 * Offsets in the buffer are 0x100 less than the equivalent in ROM.
-	 */
-
-	uint8_t header[0x50];
-
-	if (fseek(rom, 0x100, SEEK_SET) != 0)
-		err(1, "Could not locate ROM header");
-	if (fread(header, sizeof(uint8_t), sizeof(header), rom)
-	    != sizeof(header))
-		err(1, "Could not read ROM header");
-
-	if (fixlogo || trashlogo) {
-		/*
-		 * Offset 0x104–0x133: Nintendo Logo
-		 * This is a bitmap image that displays when the Game Boy is
-		 * turned on. It must be intact, or the game will not boot.
-		 */
-
-		/*
-		 * See also: global checksums at 0x14D–0x14F, They must
-		 * also be correct for the game to boot, so we fix them
-		 * as well when requested with the -f flag.
-		 */
-
-		uint8_t ninlogo[48] = {
-			0xCE, 0xED, 0x66, 0x66, 0xCC, 0x0D, 0x00, 0x0B,
-			0x03, 0x73, 0x00, 0x83, 0x00, 0x0C, 0x00, 0x0D,
-			0x00, 0x08, 0x11, 0x1F, 0x88, 0x89, 0x00, 0x0E,
-			0xDC, 0xCC, 0x6E, 0xE6, 0xDD, 0xDD, 0xD9, 0x99,
-			0xBB, 0xBB, 0x67, 0x63, 0x6E, 0x0E, 0xEC, 0xCC,
-			0xDD, 0xDC, 0x99, 0x9F, 0xBB, 0xB9, 0x33, 0x3E
-		};
-
-		if (trashlogo) {
-			for (int i = 0; i < sizeof(ninlogo); i++)
-				ninlogo[i] = ~ninlogo[i];
-		}
+	va_start(ap, fmt);
+	vfprintf(stderr, fmt, ap);
+	va_end(ap);
 
-		memcpy(header + 0x04, ninlogo, sizeof(ninlogo));
-	}
+	if (nbErrors != UINT8_MAX)
+		nbErrors++;
+}
 
-	if (settitle) {
-		/*
-		 * Offset 0x134–0x143: Game Title
-		 * This is a sixteen-character game title in ASCII (no high-
-		 * bit characters).
-		 */
-
-		/*
-		 * See also: CGB flag at 0x143. The sixteenth character of
-		 * the title is co-opted for use as the CGB flag, so they
-		 * may conflict.
-		 */
-
-		/*
-		 * See also: Game ID at 0x13F–0x142. These four ASCII
-		 * characters may conflict with the title.
-		 */
-
-		strncpy((char *)header + 0x34, title, 16);
-	}
+static const uint8_t ninLogo[] = {
+	0xCE, 0xED, 0x66, 0x66, 0xCC, 0x0D, 0x00, 0x0B,
+	0x03, 0x73, 0x00, 0x83, 0x00, 0x0C, 0x00, 0x0D,
+	0x00, 0x08, 0x11, 0x1F, 0x88, 0x89, 0x00, 0x0E,
+	0xDC, 0xCC, 0x6E, 0xE6, 0xDD, 0xDD, 0xD9, 0x99,
+	0xBB, 0xBB, 0x67, 0x63, 0x6E, 0x0E, 0xEC, 0xCC,
+	0xDD, 0xDC, 0x99, 0x9F, 0xBB, 0xB9, 0x33, 0x3E
+};
 
-	if (setid) {
-		/*
-		 * Offset 0x13F–0x142: Game ID
-		 * This is a four-character game ID in ASCII (no high-bit
-		 * characters).
-		 */
+static enum { DMG, BOTH, CGB } model = DMG; // If DMG, byte is left alone
+#define   FIX_LOGO        0x80
+#define TRASH_LOGO        0x40
+#define   FIX_HEADER_SUM  0x20
+#define TRASH_HEADER_SUM  0x10
+#define   FIX_GLOBAL_SUM  0x08
+#define TRASH_GLOBAL_SUM  0x04
+static uint8_t fixSpec = 0;
+static const char *gameID = NULL;
+static uint8_t gameIDLen;
+static bool japanese = true;
+static const char *newLicensee = NULL;
+static uint8_t newLicenseeLen;
+static uint16_t oldLicensee = 0x100;
```

Comment:
0x100 as magic default value, might add a constant for that and use it instead as "unset default" makes the check more readable that checks for 0x100 before using the value.